### PR TITLE
source code debugging

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: src/app/sourceHighlighter.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    } else {
+      if (self.currentRange) {
+        self.currentMarker = self.editor.addMarker(self.currentRange, 'highlightcode')
+      }
+    }
+  })
+
+  /* update compilation data */
+  compilerEvent.register('compilationFinished', this, function (success, data, source) {
+    if (!self.isDebugging) {
```
Comment (chriseth) [inline src/app/sourceHighlighter.js]: Do we really want this check? Don't we usually start debugging _after_ the compilation finished?

File: src/app/sourceHighlighter.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      self.reset()
+      self.compilationData = success ? data : null
```
Comment (chriseth) [inline src/app/sourceHighlighter.js]: Is it possible to store only part of the result data?

File: src/app/sourceHighlighter.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            self.sourceMappingDecoder.getLineColumnPosition(rawPos, self.currentLineColumnLayout, function (error, pos) {
+              if (!error) {
+                self.highlight(pos, rawPos.file)
+              }
+            })
+          }
+        })
+      } else {
+        self.sourceMappingDecoder.getLineColumnPosition(rawPos, self.currentLineColumnLayout, function (error, pos) {
+          if (!error) {
```
Comment (chriseth) [inline src/app/sourceHighlighter.js]: If there was an error, we should not just keep the previous highlight range.

File: src/app/sourceHighlighter.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    * highlight the statement with the given @arg position
+    *
+    * @param {Object} position - the position to highlight { start: {line, column}, end: {line, column} }
+*/
+SourceHighlighter.prototype.highlight = function (position, fileIndex) {
+  var name = utils.fileNameFromKey(this.editor.getCacheFile()) // current opened tab
+  var source = this.compilationData.sourceList[parseInt(fileIndex)] // auto switch to that tab
+  this.removeCurrentMarker()
+  if (name !== source) {
+    this.switchToFile(source) // command the app to swicth to the curent file
```
Comment (chriseth) [inline src/app/sourceHighlighter.js]: Will that trigger a recompile?

File: src/app/compiler.js [hunk 5]
```diff
@@ -119,7 +123,11 @@ function Compiler (editor, queryParams, handleGithubCall, updateFiles) {
     if (missingInputs !== undefined && missingInputs.length > 0) {
       compile(missingInputs);
     } else if (noFatalErrors) {
-      self.event.trigger('compilationFinished', [true, data, editor.getValue()]);
+      self.lastCompilationResult = {
+        data: data,
+        source: source
+      }
+      self.event.trigger('compilationFinished', [true, data, source]);
```
Comment (chriseth) [inline src/app/compiler.js]: Is the event triggered twice if there were only warnings?

File: src/app/debugger.js [hunk 6]
```diff
@@ -1,39 +1,132 @@
...
+  // register selected code item, highlight the corresponding source location
+  this.debugger.codeManager.register('changed', this, function (code, address, index) {
+    this.debugger.sourceLocationTracker.getSourceLocation(address, index, self.lastCompilationResult.data.contracts, function (error, rawLocation) {
+      if (!error) {
+        if (!self.cache.lineBreakPositionsByContent[address]) {
+          self.cache.lineBreakPositionsByContent[address] = self.sourceMappingDecoder.getLinebreakPositions(self.editor.getFile(self.lastCompilationResult.data.sourceList[rawLocation.file]))  
+        }
+        var lineColumnPos = self.sourceMappingDecoder.convertOffsetToLineColumn(rawLocation, self.cache.lineBreakPositionsByContent[address])
+        self.highlight(lineColumnPos, rawLocation);
+      }
```
Comment (chriseth) [inline src/app/debugger.js]: Unset highlight if there was an error?

File: src/app/debugger.js [hunk 7]
```diff
@@ -1,39 +1,132 @@
...
 };
 
+/**
+ * highlight the given @arg lineColumnPos
+ *
+ * @param {Object} lineColumnPos - position of the source code to hightlight {start: {line, column}, end: {line, column}}
+ * @param {Object} rawLocation - raw position of the source code to hightlight {start, length, file, jump}
+ */
+Debugger.prototype.highlight = function (lineColumnPos, rawLocation) {
+  var name = utils.fileNameFromKey(this.editor.getCacheFile()); // current opened tab
```
Comment (chriseth) [inline src/app/debugger.js]: Could also be a different file, but I think it's better to do that in a separate PR because we first need the ability to switch tabs without triggering recompilation.

File: src/app/ui-helper.js [hunk 8]
```diff
@@ -146,11 +146,12 @@ module.exports = {
       details.append(this.tableRow('Runtime Bytecode', contract.runtimeBytecode));
     }
 
+    /*
```
Comment (chriseth) [inline src/app/ui-helper.js]: This is still useful information.

File: src/app/debugger.js (hunk 7 above)
Comment (yann300) [inline src/app/debugger.js]: can we still keep that for this PR and improve this in another PR?

File: src/app/compiler.js (hunk 5 above)
Comment (yann300) [inline src/app/compiler.js]: no it is no triggered twice

File: src/app/debugger.js (hunk 7 above)
Comment (chriseth) [inline src/app/debugger.js]: yes

File: src/app/debugger.js [hunk 9]
```diff
@@ -1,39 +1,135 @@
...
+
+  this.editor.onChangeSetup(function () {
+    if (arguments.length > 0) { // if arguments.length === 0 this is a session change, we don't want to stop debugging in that case
+      self.debugger.unLoad()
+    }
+  });
+
+  // register selected code item, highlight the corresponding source location
+  this.debugger.codeManager.register('changed', this, function (code, address, index) {
+    this.debugger.sourceLocationTracker.getSourceLocation(address, index, self.lastCompilationResult.data.contracts, function (error, rawLocation) {
```
Comment (chriseth) [inline src/app/debugger.js]: This works for now, but in the future, we should not just use "last compilation results". This does not work if you want to debug a past transaction, for example.

File: src/app/compiler.js (hunk 5 above)
Comment (chriseth) [inline src/app/compiler.js]: Why? If there is an `errors` property, but it only contains warnings, then the event is triggered in line 115 and in this line.

**Comment by yann300** [comment]
contains fix for https://github.com/ethereum/browser-solidity/issues/185

**Comment by chriseth** [comment]
Please merge as soon as the functionality from remix is available in npm.
